lib/dockerregistry/filestore.go

```diff
@@ -0,0 +1,408 @@
+		if destFile == nil {
+			destFile = &syncFileInfo{}
+			destFile.RelativePath = sourceFile.RelativePath
+			destFile.AbsolutePath = joinFilepath(p.Dest, sourceFile.RelativePath)
+			destFile.filestore = destFilestore
+			ops = append(ops, &copyFileOp{
+				Source:       sourceFile,
+				Dest:         destFile,
+				ManifestFile: f,
+			})
+			continue
+		}
+
+		changed := false
+		if destFile.MD5 != sourceFile.MD5 {
+			klog.Warningf("MD5 mismatch on source %q vs dest %q: %q vs %q",
+				sourceFile.AbsolutePath,
+				destFile.AbsolutePath,
+				sourceFile.MD5,
+				destFile.MD5)
+			changed = true
+		}
+
+		if destFile.Size != sourceFile.Size {
+			klog.Warningf("Size mismatch on source %q vs dest %q: %d vs %d",
+				sourceFile.AbsolutePath,
+				destFile.AbsolutePath,
+				sourceFile.Size,
+				destFile.Size)
+			changed = true
+		}
+
+		if !changed {
+			klog.V(2).Infof("metadata match for %q", destFile.AbsolutePath)
+			continue
+		}
+		ops = append(ops, &copyFileOp{
+			Source:       sourceFile,
+			Dest:         destFile,
+			ManifestFile: f,
+		})
+	}
+
+	return ops, nil
+}
+
+func joinFilepath(filestore *Filestore, relativePath string) string {
+	s := strings.TrimSuffix(filestore.Base, "/")
+	s += "/"
+	s += strings.TrimPrefix(relativePath, "/")
+	return s
+}
+
+// Promote copies files from the Source Filestore to the Dest Filestore
+// If DryRun is set, it merely logs the operations
+func (p *FilestorePromoter) Promote(ctx context.Context) ([]syncFileOp, error) {
+	out := p.Out
+	if out == nil {
+		out = os.Stdout
+	}
+
+	sourceFilestore, err := openFilestore(ctx, p.Source, p.UseServiceAccount)
+	if err != nil {
+		return nil, err
+	}
+	destFilestore, err := openFilestore(ctx, p.Dest, p.UseServiceAccount)
+	if err != nil {
+		return nil, err
+	}
+
+	sourceFiles, err := sourceFilestore.ListFiles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	destFiles, err := destFilestore.ListFiles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	ops, err := p.computeNeededOperations(sourceFiles, destFiles, destFilestore)
+	if err != nil {
+		return nil, err
+	}
+
+	var done []syncFileOp
+	for _, op := range ops {
+		if p.DryRun {
+			if _, err := fmt.Fprintf(out, "%v\n", op); err != nil {
+				return nil, err
+			}
+		} else {
+			klog.Infof("running: %v", op)
+			if err := op.Run(ctx); err != nil {
+				return nil, err
+			}
+			done = append(done, op)
+		}
+	}
+
+	return done, nil
+}
+
+// copyFileOp manages copying a single file
+type copyFileOp struct {
+	Source *syncFileInfo
+	Dest   *syncFileInfo
+
+	ManifestFile *File
+}
+
+// Run implements syncFileOp.Run
+func (o *copyFileOp) Run(ctx context.Context) error {
+	f, err := ioutil.TempFile("", "promoter")
+	if err != nil {
+		return fmt.Errorf("error creating temp file: %v", err)
+	}
+	tempFilename := f.Name()
+	defer func() {
+		if err := os.Remove(tempFilename); err != nil {
+			klog.Warningf("unable to remove temp file %q: %v", tempFilename, err)
+		}
+	}()
+
+	// Download to our temp file
```

Comment:
Ouch --- is there no way of promoting files server-side? That's how we promote images, surely there is a way to do the same with files...

From `gsutil help cp` I found:

```
COPYING IN THE CLOUD AND METADATA PRESERVATION
  If both the source and destination URL are cloud URLs from the same
  provider, gsutil copies data "in the cloud" (i.e., without downloading
  to and uploading from the machine where you run gsutil). In addition to
  the performance and cost advantages of doing this, copying in the cloud
  preserves metadata (like Content-Type and Cache-Control). In contrast,
  when you download data from the cloud it ends up in a file, which has
  no associated metadata. Thus, unless you have some way to hold on to
  or re-create that metadata, downloading to a file will not retain the
  metadata.
```

I *strongly* support the use of `gsutil` over downloading to local disk and re-uploading. We already have examples of shelling out to `gcloud` in this project so it's not a big stretch for shelling out to `gsutil` (or even better, just using the raw API calls, if they exist, for copying files server-side from one location to another).

It would also make the tests much easier to write, I think (no need to fake files on disk...).

Reply:
I guess we have to download to perform sha256 validation (because S3/GCS do not store this info themselves)? If so, then I think what you are proposing here makes sense as we should use a stronger hash than MD5.

I agree that hashing the compressed content is not a great idea --- *unless* we also perform the compression ourselves in a deterministic manner (which I am not even sure is a thing).